fuzz: Add harness for p2p headers sync

marcofleon opened the pull request.

achow101 committed:
wallet, interfaces: Include database format in listWalletDir

achow101 committed:
gui: Use menu for wallet migration

Once legacy wallets can no longer be loaded, we need to be able to
migrate them without loading. Thus we should use a menu that lists the
wallets in the wallet directory instead of an action which migrates the
currently loaded wallet.

achow101 committed:
gui: Consolidate wallet display name to GUIUtil function

Instead of having the code for the wallet display name being copy and
pasted, use a GUIUtil function to get that for us.

achow101 committed:
gui: Use wallet name for wallet migration rather than WalletModel

To prepare for migrating wallets that are not loaded, when migration
occurs in the GUI, it should not rely on a WalletModel existing.

Co-authored-by: furszy <matiasfurszyfer@protonmail.com>

Sjors committed:
chainparams: add mainnet assumeutxo param at height 840_000

furszy committed:
gui: don't remove wallet manually before migration

marcofleon commented:
Have some comments in there still from previous code, for informational purposes for now.

marcofleon commented on src/test/fuzz/p2p_headers_sync.cpp:
```diff
@@ -0,0 +1,202 @@
+CBlock ConsumeBlock(FuzzedDataProvider& fuzzed_data_provider, const uint256& prev_hash, uint32_t prev_nbits)
+{
+    auto header = ConsumeHeader(fuzzed_data_provider, prev_hash, prev_nbits);
+    // Doesn't need to be a valid block but it needs to have one transaction
```
This comment should probably be changed. It's still true that it doesn't need to be a valid block correct? But it does need to be unmutated... not sure what the actual term would be.

dergoegge replied:
The comment should describe why we construct the block this way, e.g. we want it to reach the headers acceptance logic when processing a block message and for that it needs to pass the mutation checks.

marcofleon replied:
done

theStack committed:
doc: add missing "testnet4" network string in RPC/init help texts

The following bitcoind parameters / RPC calls missed the "testnet4"
network string:
    - `-chain=` parameter
    - `getblockchaininfo` RPC, "chain" result
    - `getmininginfo` RPC, "chain" result

dergoegge commented on src/net_processing.h:
```diff
@@ -71,6 +74,8 @@ class PeerManager : public CValidationInterface, public NetEventsInterface
         //! a test-only option).
         bool deterministic_rng{false};
+        //! Number of headers sent in one getheaders message result.
+        uint32_t max_headers_result{MAX_HEADERS_RESULTS};
```
Better to split the peer manager changes out into a separate commit

marcofleon replied:
done

dergoegge commented on src/pow.cpp:
```diff
@@ -134,9 +134,24 @@ bool PermittedDifficultyTransition(const Consensus::Params& params, int64_t heig
 bool CheckProofOfWork(uint256 hash, unsigned int nBits, const Consensus::Params& params)
 {
-    bool fNegative;
+#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
```
We still need a separate `CheckProofOfWorkImpl` that can be fuzzed on its own.

Otherwise, https://github.com/bitcoin/bitcoin/blob/5fdbc8b4ee609d7f58dd4dff30e506a7400cd768/src/test/fuzz/pow.cpp#L83 would not be testing anything.

marcofleon replied:
done

marcofleon committed:
net_processing: Make MAX_HEADERS_RESULTS a PeerManager option

marcofleon committed:
Add FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION in PoW check

To avoid PoW being a blocker for fuzz tests,
`FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION` is used in fuzz builds to
bypass the actual PoW validation in `CheckProofOfWork`. It's
replaced with a check on the last byte of the hash, which allows the
fuzzer to quickly generate (in)valid blocks by checking a single bit,
rather than performing the full PoW computation.

If PoW is the target of a fuzz test, then it should call
`CheckProofOfWorkImpl`.

marcofleon committed:
fuzz: Add harness for p2p headers sync

marcofleon commented:
@dergoegge I think this is good to go. Let me know what you think. I separated into 3 commits, makes the most sense.

dergoegge commented:
I think commit https://github.com/marcofleon/bitcoin/pull/1/commits/2e4f5f0ee7f447ea2e966269a744928ae73e415f should come before https://github.com/marcofleon/bitcoin/pull/1/commits/8c2b1d90d9f73720ceff080fdb38bec1230cf0c0.

You could also consider formatting the commit message in https://github.com/marcofleon/bitcoin/pull/1/commits/2e4f5f0ee7f447ea2e966269a744928ae73e415f with e.g. "gq" in vim, so that the lines wrap nicely.

marcofleon committed:
build: Automatically define FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION for fuzz builds

glozow committed:
Merge bitcoin/bitcoin#30642: doc: add missing "testnet4" network string in RPC/init help texts

701530045553f2b9671a3fffea301bf4dc954514 doc: add missing "testnet4" network string in RPC/init help texts (Sebastian Falbesoner)

Pull request description:

  The following bitcoind parameters / RPC calls still missed the "testnet4" network string:
      - `-chain=` parameter
      - `getblockchaininfo` RPC, `"chain"` result
      - `getmininginfo` RPC, `"chain"` result

  The occurences were found via `$ git grep \".*main.*test.*\"`.

ACKs for top commit:
  maflcko:
    review ACK 701530045553f2b9671a3fffea301bf4dc954514
  glozow:
    ACK 701530045553f2b9671a3fffea301bf4dc954514
  tdb3:
    ACK 701530045553f2b9671a3fffea301bf4dc954514
  BrandonOdiwuor:
    ACK 701530045553f2b9671a3fffea301bf4dc954514
  danielabrozzoni:
    ACK 701530045553f2b9671a3fffea301bf4dc954514

Tree-SHA512: 99bf5c2b4cf28651feaff2fc7d4669961012dfa8379d8522251540ae1b8fc77d1761b75395903b527580530f42a3c1fd2dd2d8dba4ffbc9b6e55cb357c3a271b

martinsaposnic committed:
Deduplicate list of chain strings in RPC help texts

hebasto committed:
Merge bitcoin-core/gui#824: Migrate legacy wallets that are not loaded

8f2522d242961ceb9e79672aa43e856863a1a6dd gui: Use menu for wallet migration (Ava Chow)
d56a450bf5172e2c3f4b9a2786e71268019e1277 gui: Use wallet name for wallet migration rather than WalletModel (Ava Chow)
c3918583dd5fcd9001136da2192e02e092128901 gui: don't remove wallet manually before migration (furszy)
bfba63880fbb1108b73540faeb0620ba24b8cdd0 gui: Consolidate wallet display name to GUIUtil function (Ava Chow)
28fc562f2692af4f37f918d4ae31c4d115e03aee wallet, interfaces: Include database format in listWalletDir (Ava Chow)

Pull request description:

  Currently the Migrate Wallet menu item can only be used to migrate the currently loaded wallet. This is not suitable for the future when legacy wallets can no longer be loaded at all, but should still be able to be migrated. This PR changes that menu item into a menu list like Open Wallet and lets users migrate any legacy wallet in their wallet directory regardless of the wallets loaded.

  One issue I ran into was dealing with encrypted wallets. Ideally, we would detect whether a wallet is encrypted, and prompt the user for their passphrase at that time. However, that's actually difficult to do in the GUI since migration will unload the wallet if it was already loaded, and reload it without connecting it to any signals or interfaces. Only then can it detect whether a wallet is encrypted, but then there is no `WalletModel` or even an `interfaces::Wallet` that the GUI could use to unlock it via a callback.

  To deal with this, I've opted to just add a button to the migration dialog box that has the user enter their passphrase first, along with instructional text to use that button if their wallet was encrypted. If the user enters the wrong passphrase or clicked the other button that does not prompt for the passphrase, migration will fail with a message indicating that the passphrase was incorrect.

ACKs for top commit:
  hebasto:
    ACK 8f2522d242961ceb9e79672aa43e856863a1a6dd.
  furszy:
    ACK 8f2522d

Tree-SHA512: a0e3b70dbfcacb89617956510ebcea94cad8617a987c68fe39fa16ac1721190b7cf7afc156c39b9032920cfb67b5d4ca28791681f5021d92d16acc691387afa1

dergoegge commented:
Looks good to me, will do a deeper review on the real PR.

nit on the title: this is a harness for the headers pre-sync.

hebasto committed:
qt: Update translation source file for v28.0 string freeze

The diff is produced by running `make -C src translate`.

hebasto committed:
Merge bitcoin-core/gui#833: Update translation source file for v28.0 string freeze

770b0348c0abe2f63b56cc78c7b7a0e6b4057ce1 qt: Update translation source file for v28.0 string freeze (Hennadii Stepanov)

Pull request description:

  This PR updates the `src/qt/locale/bitcoin_en.xlf` translation source file according to the [Release schedule for 28.0](https://github.com/bitcoin/bitcoin/issues/29891).

  Note for reviewers: it is expected to get a zero diff after running `make -C src translate` locally.

ACKs for top commit:
  stickies-v:
    re-ACK 770b0348c0abe2f63b56cc78c7b7a0e6b4057ce1
  pablomartin4btc:
    re-ACK 770b0348c0abe2f63b56cc78c7b7a0e6b4057ce1

Tree-SHA512: 11dd26c470411aefc2e4f897c605162027a00e2a0ab1dcec9a1784c053349a3feaeedda7b649476ff528231801629e0ef342a48430ef54a4ec75ac1548c56d4f

glozow committed:
[doc] mention bip94 support

achow101 committed:
Merge bitcoin/bitcoin#30648: doc: Deduplicate list of possible chain strings in RPC help texts

9b297555207b4ea54bc0051f09c7084797aa9def Deduplicate list of chain strings in RPC help texts (Martin Saposnic)

Pull request description:

  As mentioned in issue https://github.com/bitcoin/bitcoin/issues/30645:

  Many command line parameter and RPC help texts currently contain the list of chain/network names hardcoded ("main, test, testnet4, signet, regtest"), which is error-prone as it can easily happen to miss an instance if the list ever changes again.

  This PR deduplicates the list of possible chain/network strings in RPC/parameter help texts, and it creates a macro `LIST_CHAIN_NAMES` in src/chainparamsbase.h. In the future, there is only 1 place where that list of possible values lives, so maintainability is improved and errors are avoided.

  All three places where this change impacts:

  ```
  ./bitcoin-cli --help
  ./bitcoin-cli help getblockchaininfo
  ./bitcoin-cli help getmininginfo
  ```

  They all return the correct string `"main, test, testnet4, signet, regtest"`

  See https://github.com/bitcoin/bitcoin/pull/30642#discussion_r1714711575

ACKs for top commit:
  maflcko:
    lgtm ACK 9b297555207b4ea54bc0051f09c7084797aa9def
  achow101:
    ACK 9b297555207b4ea54bc0051f09c7084797aa9def
  MarnixCroes:
    ACK 9b297555207b4ea54bc0051f09c7084797aa9def
  theStack:
    ACK 9b297555207b4ea54bc0051f09c7084797aa9def
  danielabrozzoni:
    ACK 9b297555207b4ea54bc0051f09c7084797aa9def

Tree-SHA512: 1e961bcbe40b0f17a87a2437eb4ba1bb89468fd1b5a39599d72a00ef75cb4009e7d2f05d0a621bb904fecf681c55b8a219fcfe4d44d5d27f27cdda20882b1323

achow101 committed:
Merge bitcoin/bitcoin#28553: validation: assumeutxo params mainnet

1610643c8b37a9f674b236cfa79abf8f8aaf1410 chainparams: add mainnet assumeutxo param at height 840_000 (Sjors Provoost)

Pull request description:

  This adds snapshot parameters for mainnet block 840,000.

  You can generate the snapshot yourself using `./contrib/devtools/utxo_snapshot.sh` or download my torrent:
  * torrent: `magnet:?xt=urn:btih:596c26cc709e213fdfec997183ff67067241440c&dn=utxo-840000.dat&tr=udp%3A%2F%2Ftracker.bitcoin.sprovoost.nl%3A6969`

  It would be a good idea to test:
  1. That you can produce the same snapshot file, sha256 sum:

  ```
  dc4bb43d58d6a25e91eae93eb052d72e3318bd98ec62a5d0c11817cefbba177b utxo-840000.dat
  ```

  2. That the snapshot works

ACKs for top commit:
  fjahr:
    re-ACK 1610643c8b37a9f674b236cfa79abf8f8aaf1410
  achow101:
    ACK 1610643c8b37a9f674b236cfa79abf8f8aaf1410
  theStack:
    Tested ACK 1610643c8b37a9f674b236cfa79abf8f8aaf1410
  mzumsande:
    tested ACK 1610643c8b37a9f674b236cfa79abf8f8aaf1410
  willcl-ark:
    tACK 1610643c8b37a9f674b236cfa79abf8f8aaf1410

Tree-SHA512: 581d8e86379bb044324f04f8559dd0a8946b6e2b145d5f25b38727b30b8cf13d6ac3c8777ff06554d3cf1a072809f7b5fbd693239868578f25dceafe5ba5f57c

glozow committed:
Merge bitcoin/bitcoin#30655: doc: mention bip94 support

99eeb51bf65fa00ee863f32d70f478bb9db0e351 [doc] mention bip94 support (glozow)

Pull request description:

  Followup to #29775, noticed while looking at #30604 and #30647. See [release process](https://github.com/bitcoin/bitcoin/blob/master/doc/release-process.md#before-every-major-and-minor-release).

ACKs for top commit:
  maflcko:
    ACK 99eeb51bf65fa00ee863f32d70f478bb9db0e351
  fjahr:
    ACK 99eeb51bf65fa00ee863f32d70f478bb9db0e351
  tdb3:
    ACK 99eeb51bf65fa00ee863f32d70f478bb9db0e351

Tree-SHA512: 95838d3ace7e5d7b1a2481f2d7bd82902081713e6e89dbf21e0dad16d1cf5295e0c1cfda1f03af72304a5844743d24769f5fa04d4dc9f02f36462ef0ae82a552
